File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.js
function solution(cards) {
    const dfs = (idx, visited, count) => {
        if (visited[idx]) return count;
        visited[idx] = true;
        return dfs(cards[idx - 1], visited, count + 1);
    };

    let maxScore = 0;

    for (let i = 0; i < cards.length; i++) {
        const visited = new Array(cards.length + 1).fill(false);
        const score1 = dfs(cards[i], visited, 0); // 4

        const leftCards = cards.filter((_, index) => !visited[index + 1]); // [2,5,6]
        const visited2 = new Array(leftCards.length + 1).fill(false);
        for (j = 0; j < leftCards.length; j++) {
            const score2 = leftCards.length === 0 ? 0 : dfs(leftCards[j], visited2, 0);
            const score = score1 * score2;
            maxScore = Math.max(maxScore, score);
        }
    }

    return maxScore;
}

File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.py
def solution(cards):
    group_sizes = []
    queue = []
    visited = [False] * len(cards)
    cards = list(map(lambda x: x - 1, cards)) # cards는 1 based index이므로 0 based index로 수정해준다.
    
    for i in range(len(cards)):        
        if not visited[i]:
            queue.append(i) # BFS 시작점 세팅
            visited[i] = True
            count = 1
    
            while queue: # [0]
                cur_idx = queue.pop(0) # 인덱스 0번째 상자에 있는
                cur_box = cards[cur_idx] # 8번 카드
                if not visited[cur_box]:
                    queue.append(cur_box)
                    visited[cur_box] = True
                    count += 1
            # BFS 종료 후
            group_sizes.append(count)
    
    group_sizes.sort(reverse = True)
    
    return 0 if len(group_sizes) <= 1 else group_sizes[0] * group_sizes[1]